api/v1alpha1/virtualmachine_conversion.go

```diff
@@ -892,13 +892,150 @@ func restore_v1alpha2_VirtualMachineReadinessProbeSpec(
+		// Order of evaluation of objects for v1a1 PreReq Condition:
+		// 1. VM Class then
+		// 2. VM Image
+
+		if preReqCond.Reason == VirtualMachineClassNotFoundReason {
+			conditions = append(conditions, metav1.Condition{
+				Type:               v1alpha2.VirtualMachineConditionClassReady,
+				Status:             metav1.ConditionFalse,
+				LastTransitionTime: preReqCond.LastTransitionTime,
+				Reason:             string(metav1.ConditionFalse),
+			})
+			// v1a1 Image preReq hasn't been evaluated yet -- mark unknown
+			conditions = append(conditions, metav1.Condition{
+				Type:               v1alpha2.VirtualMachineConditionImageReady,
+				Status:             metav1.ConditionUnknown,
+				LastTransitionTime: preReqCond.LastTransitionTime,
+				Reason:             string(metav1.ConditionUnknown),
+			})
+		} else if preReqCond.Reason == VirtualMachineImageNotFoundReason {
+			conditions = append(conditions, metav1.Condition{
+				Type:               v1alpha2.VirtualMachineConditionClassReady,
+				Status:             metav1.ConditionTrue,
+				LastTransitionTime: preReqCond.LastTransitionTime,
+				Reason:             string(metav1.ConditionTrue),
+			})
+			conditions = append(conditions, metav1.Condition{
+				Type:               v1alpha2.VirtualMachineConditionImageReady,
+				Status:             metav1.ConditionFalse,
+				LastTransitionTime: preReqCond.LastTransitionTime,
+				Reason:             string(metav1.ConditionFalse),
+			})
+		}
+
+		// v1a1 setPolicy has not been evaluated when preReq condition is false. Mark unknown in v1a2 when
+		// resourcePolicy is specified in VM Spec.
+		if dst.Spec.Reserved != nil && dst.Spec.Reserved.ResourcePolicyName != "" {
+			conditions = append(conditions, metav1.Condition{
+				Type:               v1alpha2.VirtualMachineConditionVMSetResourcePolicyReady,
+				Status:             metav1.ConditionUnknown,
+				LastTransitionTime: preReqCond.LastTransitionTime,
+				Reason:             string(metav1.ConditionUnknown),
+			})
+		}
+
+		// v1a1 vmMetadata has not been evaluated when preReq condition is false. Mark unknown in v1a2 when
+		// bootstrap is specified in VM Spec.
+		if dst.Spec.Bootstrap != nil {
+			conditions = append(conditions, metav1.Condition{
+				Type:               v1alpha2.VirtualMachineConditionBootstrapReady,
+				Status:             metav1.ConditionUnknown,
+				LastTransitionTime: preReqCond.LastTransitionTime,
+				Reason:             string(metav1.ConditionUnknown),
+			})
+		}
+	}
+
+	return append(dst.Status.Conditions, conditions...)
+}
+
 // ConvertTo converts this VirtualMachine to the Hub version.
 func (src *VirtualMachine) ConvertTo(dstRaw conversion.Hub) error {
 	dst := dstRaw.(*v1alpha2.VirtualMachine)
 	if err := Convert_v1alpha1_VirtualMachine_To_v1alpha2_VirtualMachine(src, dst, nil); err != nil {
 		return err
 	}
 
+	newConditions := convert_v1alpha1_PreReqsReadyCondition_to_v1alpha2_Conditions(dst)
```

Comment:
Can we move this to L1042 instead? If we're able to restore from the saved annotation, we know that the PreReq won't be there (and L1049 unconditionally restores that Status anyways)

Reply:
Basically only do this `convert_v1alpha1_PreReqsReadyCondition` when there is nothing to restore from the conversion annotation?.